Delaunay Core Bifiltration

odinhg opened the pull request:
This pull request includes the addition of the **Delaunay core bifiltration** and a minor fix of the three annulus synthetic dataset.

### Delaunay Core Bifiltration

I implemented the function `CoreDelaunay` in the file `multipers/filtration/filtrations.py`. The imports `sqrt` from `math` and `KDTree` from `scipy.spatial` are added to this file. I also added some simple tests for `CoreDelaunay` in `multipers/tests/test_filtrations.py`. Here, I imported `gudi` to do a comparison against the alpha complex.

### Minor Fix of `three_annulus` in `multipers/data/synthetic.py`

The changes ensure that the function returns a point cloud with `n1 + n2` points, even if `n1` is not divisible by three. Previously, calling `three_annulus(100, 50)` would return a point cloud with 149 points, which, I think, is a bit unintuitive.

### Questions

**Formatting:** I automatically ran `ruff format` on the Python files without thinking. This changed the style of some existing code. Please let me know if I should revert these formatting changes. Also, is there a preferred formatting style I should follow for Python code in `multipers`?

**Docstrings:** I tried to imitate the docstring style already used in `multipers/filtration/filtrations.py`. Does it look okay, or should I change it?

DavidLapous commented on multipers/filtration/filtrations.py:
```diff
@@ -140,16 +164,105 @@ def Cubical(image:ArrayLike, **slicer_kwargs):
+    knn_distances = KDTree(points).query(points, k=ks)[0]
+
+    # Group simplices by dimension
+    simplices_in_dimension = {dim: [] for dim in range(alpha_complex.dimension() + 1)}
+    for simplex, alpha_sq in alpha_complex.get_simplices():
+        alpha = sqrt(alpha_sq)
+        simplices_in_dimension[len(simplex) - 1].append((simplex, alpha))
```
This may be optimizable:
 - as the dimensions will be some kind of range, a tuple may be faster than this dictionary
 - the `sqrt` can be applied afterwards with a vectorized `np.sqrt`
 - dimension-wise simplices and filtrations can be put in arrays

But since we planned to optimize the code later it's not really an issue

odinhg replied:
I have now vectorized parts of the `CoreDelaunay` implementation using dimension-indexed lists of NumPy arrays for the simplices and the alpha filtration values. I also vectorized the computation of `critical_radii` so we no longer iterate over all simplices in the batch insertion step.

I didn't run any benchmarking experiments to compare the speed difference before and after vectorizing. Please let me know if you have any comments/suggestions to these changes.

DavidLapous reviewed (commented):
Thanks for this contribution and the tests ! This seems reasonable, I'll try to run this code tomorrow.

For your questions:
> Formatting: ...

I don't have any rule for the formatting, although this library aims to be "gudhi-compatible" (for a future merge) so if you want to use one you can use the one of Gudhi.
No need to revert the formatting, but in general, try to avoid this so that the changes are easier to read.
> Docstring: ...

It looks good !  You can also render the doc ([see here](https://davidlapous.github.io/multipers/contributions.html#building-the-documentation)) on your end and see if it looks good there. 

For the optimization part : I think doing cython will be enough here, since the cpp multicritical filtration class is directly accessible there, as well as the simplextreemulti iterators.

DavidLapous commented:
Great !  
I was also thinking about how to make this contribution more visible in the library; 
If you are willing to, you can do a jupyter notebook, with your name + link to your paper (e.g., similar to gudhi's TDA tutorial notebooks) explaining this filtration in an example. For instance, in a new "Filtrations" section above the example zoo 
https://github.com/DavidLapous/multipers/blob/51d9845147deea6a2dda48626d7f1722ba4039b6/docs/index.rst?plain=1#L118-L126
with the corresponding notebook in the `docs/notebooks` folder.

You can also add the paper in the list in the readme.
If you think of something else, don't hesitate.

odinhg commented:
I have added a tutorial notebook showing an example of Delaunay core persistence with the `three_annulus` dataset and added it to a new `Filtration` section in the documentation like you suggested.

DavidLapous reviewed (commented):
Looks good ! 

My final comments before merging :
 - there is a `wich` typo in the intro of the notebook
 - there is a (maybe hidden) little function to compute the bounding box of multipers objects, so your code to compute a bounding box can be achieved with a
  ```python
 box = mp.grids.compute_bounding_box(pers)
  ```
 - I don't know if it's better or not, but do you think (`k_max` and `k_step`) is better than directly giving the `ks = range(1,k_max,k_step)` ? I already have this `ks` interface for the landscapes for instance. 
 - In the last cell, by removing the `degree` argument, i.e.,
 ```python
 pers.plot(alpha=0.9, dpi=400, xlabel="r", ylabel="k", min_persistence=10e-3, box=box)
```
you can directly plot the two dimensions at once

Thanks for the notebook !

DavidLapous commented on multipers/filtration/filtrations.py:
```diff
@@ -147,9 +149,86 @@ def DegreeRips(*, points = None, distance_matrix=None, ks=None, threshold_radius
+        vertex_arrays_in_dimension[dim].append(simplex)
+
+    alphas_in_dimension = [np.sqrt(np.array(alpha_squared, dtype=np.float64)) for alpha_squared in squared_alphas_in_dimension]
+    vertex_arrays_in_dimension = [np.array(vertex_array, dtype=int) for vertex_array in vertex_arrays_in_dimension]
```
Small optimization, the `insert_batch` will copy this to `int32` otherwise   
```suggestion
    vertex_arrays_in_dimension = [np.array(vertex_array, dtype=np.int32) for vertex_array in vertex_arrays_in_dimension]
```

DavidLapous commented on multipers/filtration/filtrations.py:
```diff
@@ -147,9 +149,86 @@ def DegreeRips(*, points = None, distance_matrix=None, ks=None, threshold_radius
+    points: ArrayLike,
+    *,
+    beta: float = 1.0,
+    k_max: int | None = None,
```
```suggestion
    k_max: Optional[int] = None,
```

DavidLapous commented on multipers/filtration/filtrations.py:
```diff
@@ -1,5 +1,7 @@
 import gudhi as gd
 import numpy as np
+from math import sqrt
```
```suggestion
```

odinhg commented:
Thanks for the corrections and good suggestions!

- [x] Fixed the typo *"wich"* in the notebook.
- [x] Used the existing `compute_bounding_box` function from `multipers.grids` in the notebook.
- [x] Removed the `degree` argument when plotting in the notebook.
- [x] Changed `k_max` and `k_step` into a single keyword argument `ks: Optional[list|np.ndarray] = None` in `CoreDelaunay`.
- [x] Updated tutorial, tests and docstring with these changes.

DavidLapous reviewed (commented):
There is a little issue if `ks` is not an array nor a list, with a fix below.

In the notebook : 
 - there is a `verbose=True`, I don't know if you want to keep it
 - I think I prefer `radius` in the legends rather that's `r`, but that's up to you

DavidLapous commented:
Thanks ! LGTM, I can merge it now if you're fine with it

DavidLapous commented:
Thanks again for this Odin ! I will try to do a release soon, but I need to find time to do the notebook for the other filtrations first, and maybe simplify some other notebooks with this new filtration module
